pkg/reconciler/instances/istio/reset/pod/handler.go

```diff
@@ -112,10 +134,162 @@ func (i *RolloutHandler) ExecuteAndWaitFor(context context.Context, object Custo
+	switch object.Kind {
+	case "Deployment":
+		err := annotatePodsFor(context, getDeploymentLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	case "DaemonSet":
+		err := annotatePodsFor(context, getDaemonSetLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	case "ReplicaSet":
+		err := annotatePodsFor(context, getReplicaSetLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+
+	case "StatefulSet":
+		err := annotatePodsFor(context, getStatefulSetLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	case "Job":
+		err := annotatePodsFor(context, getJobLabelSelector, client, object, key, val)
+		if err != nil {
+			return err
+		}
+	default:
+		err := fmt.Errorf("kind %s not found", object.Kind)
+		return err
+	}
+	return nil
+}
+
+type labelSelectorGetter func(context context.Context, client kubernetes.Interface, object CustomObject) (string, error)
+
+func annotatePodsFor(context context.Context, f labelSelectorGetter, client kubernetes.Interface, object CustomObject, annotationKey, annotationVal string) error {
+	labelSelector, err := f(context, client, object)
+	if err != nil {
+		return err
+	}
+	podList, err := listPodsWithLabelSelector(context, client, object.Namespace, labelSelector)
+	if err != nil {
+		return err
+	}
+	err = annotatePods(context, client, podList, annotationKey, annotationVal)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func listPodsWithLabelSelector(context context.Context, client kubernetes.Interface, namespace string, labelSelector string) (*v1.PodList, error) {
+	podList, err := client.CoreV1().Pods(namespace).List(context, metav1.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return nil, err
+	}
+	return podList, nil
+}
+
+func annotatePods(context context.Context, kubeClient kubernetes.Interface, podList *v1.PodList, key string, timeout string) error {
+	for _, pod := range podList.Items {
+		if pod.Annotations == nil {
+			pod.Annotations = make(map[string]string)
+		}
+		pod.Annotations[key] = timeout
+		_, err := kubeClient.CoreV1().Pods(pod.Namespace).Update(context, &pod, metav1.UpdateOptions{}) //nolint:gosec
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func getJobLabelSelector(context context.Context, client kubernetes.Interface, object CustomObject) (string, error) {
+	job, err := client.BatchV1().Jobs(object.Namespace).Get(context, object.Name, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	labelSelector := buildLabelSelector(job.Spec.Selector.MatchLabels)
+	return labelSelector, nil
+}
+
+func getDeploymentLabelSelector(context context.Context, client kubernetes.Interface, object CustomObject) (string, error) {
+	deployment, err := client.AppsV1().Deployments(object.Namespace).Get(context, object.Name, metav1.GetOptions{})
```

Comment:
This function could be simplified to return only labels taken from `.Spec.Selector.MatchLabels` and in the layer above `buildLabelSelector(...)` could be called once for any labels coming from here. It would make this contract cleaner. What do you think?